### PR TITLE
Converting DependencyLicensesTask and UpdateShasTask to java

### PR DESCRIPTION
**Pull request by Megamiun**
Relates to #34459

### CODE REVIEW

**Comment by atorok** [comment]
@elasticmachine test this please

**Review by alpar-t — COMMENTED** [review]
Thansk for your  contribution @Megamiun .
I left some comments. Still have to look at the tests. 
Perhaps would be good to add a testkit based test to make sure that the up-to-date checks work as they should. Calling the task twice should result in the second time being up-to-date

File: buildSrc/src/main/java/org/elasticsearch/gradle/precommit/DependencyLicensesTask.java [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+
+    private final Logger logger = Logging.getLogger(getClass());
+
+    private static final String SHA_EXTENSION = ".sha1";
+
+    // TODO: we should be able to default this to eg compile deps, but we need to move the licenses
+    // check from distribution to core (ie this should only be run on java projects)
+    /** A collection of jar files that should be checked. */
+    @InputFiles
+    public FileCollection dependencies;
```
Comment (atorok) [inline buildSrc/src/main/java/org/elasticsearch/gradle/precommit/DependencyLicensesTask.java]: I think we should expose this differently,  but that can go into a separate PR. 
Assigning this from configurations makes it way too easy to resolve the configurations too early. 
We should get the dependencies on project like it's being assigned in `BuildPlugin` and offer ways to filter them, like having a `List<Predicate>` for projects that need it, like `server/build.gradle`

File: buildSrc/src/main/java/org/elasticsearch/gradle/precommit/DependencyLicensesTask.java [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    private Map<String, String> mappings = new LinkedHashMap<>();
+
+    /** Names of dependencies whose shas should not exist. */
+    private Set<String> ignoreShas = new HashSet<>();
+
+    /**
+     * Add a mapping from a regex pattern for the jar name, to a prefix to find
+     * the LICENSE and NOTICE file for that jar.
+     */
+    @Input
```
[the remark this answers is not in the document]
Comment (Megamiun) [reply buildSrc/src/main/java/org/elasticsearch/gradle/precommit/DependencyLicensesTask.java]: It seems the annotation was already in place on the groovy file. So, was it wrong before already?

File: buildSrc/src/main/java/org/elasticsearch/gradle/precommit/DependencyLicensesTask.java (hunk 1 above)
Comment (Megamiun) [reply buildSrc/src/main/java/org/elasticsearch/gradle/precommit/DependencyLicensesTask.java]: Didn't do this one yet, think I will leave it to another future PR

**Comment by Megamiun** [comment]
About the test-kit tests, I don't understand corretcly what should I do. Just a test to see if the status would be UP-TO-DATE on a repeated run?

**Comment by atorok** [comment]
You can look at tests that implement  `GradleIntegrationTestCase` for some examples. 
You have to set up a project that applies the build plugin and check that the tasks do what it's expected of them. Then run it a second time and assert that it's up to date. This will catch incorrectly wired task inputs for the most part. You could also   change some of the inputs or outputs and validate that it's not considered up to date. 

@elasticmachine test this please

**Comment by atorok** [comment]
@Megamiun you'll have to merge master to get the builds passing.

**Comment by Megamiun** [comment]
Sorry for the delay, almost finished here.

But I saw something, it seems this task can not be used as up-to-date. I put it to run with --info and received:

```
Task ':x-pack:plugin:ml:dependencyLicenses' is not up-to-date because:
  Task has not declared any outputs despite executing actions.
```

It seems that by not existing output, gradle doesn't skip execution because of up-to-date status.

Probably by tomorrow I can make the MR

**Comment by tomcallahan** [comment]
@Megamiun are you still interested in pursuing this PR ?

**Comment by Megamiun** [comment]
Yes, thanks for remembering me, will try to do this weekend.

But even so, @atorok, about the UP-TO-DATE, is it really the desired behaviour?

**Comment by rjernst** [comment]
>  about the UP-TO-DATE, is it really the desired behaviour?

Nope, we should fix this (although it doesn't have to be in this PR). The way we have done this in other cases is to use a "marker" file that we simply touch when the task completes, and configure that as the task output. For example, see ForbiddenPatternsTask.

**Comment by Megamiun** [comment]
@elasticmachine test this please

**Review by alpar-t — COMMENTED** [review]
@elasticmachine test this please

File: buildSrc/build.gradle [hunk 3]
```diff
@@ -113,7 +113,9 @@ dependencies {
   compile localGroovy()
   compile "com.carrotsearch.randomizedtesting:junit4-ant:${props.getProperty('randomizedrunner')}"
   compile "com.carrotsearch.randomizedtesting:randomizedtesting-runner:${props.getProperty('randomizedrunner')}"
-  
+
+  compile "commons-codec:commons-codec:${props.getProperty('commonscodec')}"
```
[the remark this answers is not in the document]
Comment (Megamiun) [reply buildSrc/build.gradle]: I added this to add the possibility of encoding the sha1 byte array to an hex string.

It seems the way to do it is a little more laborious in pure java and the encodeHex method was a Groovy method.

If prefered, I can change the impl to use java directly.

And when making changes, I just need to rebase my branch, correct?

[the remark this answers is not in the document]
Comment (Megamiun) [reply buildSrc/build.gradle]: Oh, and the integTest task takes some hours(it didn't complete yet), on my computer, is it normal?

[the remark this answers is not in the document]
Comment (atorok) [reply buildSrc/build.gradle]: @Megamiun thanks for the explanation, I think that's good enough reason to add it.
Unfortunately our integ tests do take that long currently, there's many of them and we are working to optimize execution to make them faster.

@elasticmachine test this please

**Comment by atorok** [comment]
@elasticmachine tesdt this please

**Comment by atorok** [comment]
@elasticmachine test this please

**Comment by atorok** [comment]
@elasticmachine test this please

**Comment by Megamiun** [comment]
@elasticmachine test this please

**Comment by Megamiun** [comment]
Hello, tried to resolve the IT tests problem, is it possible to run again the tests? Unfortunately the time it takes to tuns the tests makes it not possible to test it on my pc

**Comment by atorok** [comment]
@Megamiun you can run the test you added alone with this command: `./gradlew :build-tools:integTest --tests="*DependencyLicensesTasksIT"` 

@elasticmachine test this please

**Comment by Megamiun** [comment]
I think maybe there is a problem with my pc, it has been eleven hours and the integtests are still running. And I used the command to run opnly mine

**Comment by Megamiun** [comment]
Hello, @atarok?

File: buildSrc/src/test/java/org/elasticsearch/gradle/precommit/DependencyLicensesTasksIT.java [hunk 4]
```diff
@@ -38,24 +38,27 @@ public void testDependenciesOk() {
...
 
-        assertTaskFailed(runner.build(), ":no_dependencies_and_no_folder:dependencyLicenses");
+        assertTaskFailed(runner.buildAndFail(), ":no_dependencies_and_no_folder:dependencyLicenses");
     }
 
     public void testDependenciesMissing() {
         GradleRunner runner = getRunner("dependencies_missing");
 
-        assertTaskFailed(runner.build(), ":dependencies_missing:dependencyLicenses");
+        assertTaskFailed(runner.buildAndFail(), ":dependencies_missing:dependencyLicenses");
```
Comment (atorok) [inline buildSrc/src/test/java/org/elasticsearch/gradle/precommit/DependencyLicensesTasksIT.java]: This change is needed, but when asserting for failure we should assert for something in the output to make sure we are getting the failure we want.
I can see that from the last test run we got `Plugin with id 'elasticsearch.build' not found.` so I don't think we are getting the intended failure here yet the test would pass.

The reason for the failure is that you must use the new Gradle syntax to apply the plugin. take a look at `buildSrc/src/testKit/elasticsearch.build/build.gradle`

**Comment by Megamiun** [comment]
Sorry, @atorok, I closed this PR by error. Now I am doing one more try of running the tests. I will open a new PR once I finish this.
